### DISCUSSION TITLE
使用 R base 抓取网页数据--基于红黑统计公报库 | /project/r-crawl-data2/

### DISCUSSION THREAD

**Discussion by giscus** [earfanfan/yuanfan#639] [Announcements]
https://yuanfan.rbind.io/project/r-crawl-data2/

**Comment by CyrusYip** [comment]
> 在浏览器中打开首页，按 F12 打开“开发者模式”，使用“元素选择器”选中一个县级公报，查看对应的元素内容，如下所示。

可以直接右击后点「检查」。

---

我觉得找 HTML 元素用 CSS 选择器会比较好，正则表达式写得头痛，看着也头痛。

以 https://tjgb.hongheiku.com/category/xjtjgb 为例，用 CSS 选择器可以只选择合适的元素。

```js
const links = document.querySelectorAll(".excerpt > header > h2 > a")
links.forEach(link => { console.log(`${link.title}: ${link.href}`) })
```

![Screenshot_20241023_233801](https://github.com/user-attachments/assets/dbd54019-ce7e-4fd1-b2b6-1276b74ace03)

R 语言应该有库可以做这样的操作吧。

**Reply by yihui** [reply]
可以 xml2 包配合 XPath 去选择。XPath 和 CSS 选择器有类似之处。不过我自己在大多数情况下都是正则表达式处理 HTML，这种方式不如 XPath 严谨，但如果 HTML 比较规整的话，正则也够了。

**Reply by earfanfan** [reply]
(⊙o⊙)…原来可以直接右键选“检查”，又学到了。

是可以用 xml2 包，我第一次尝试抓取<https://yuanfan.rbind.io/project/r-crawl-data/>就试过的。不过这次鼓捣的时候，抓一个的情况都是在公司干活的空隙摸鱼鼓捣的，而公司电脑不能正常安装 xml2 包（PS那又是一个令人颇感无奈的故事了），所以全都是用 R base 鼓捣的。其实之前我也是觉得用现成的包现成的函数直接看着 XPath 选要方便得多，不过刚好最近写 R base 字符串处理函数又学了正则表达式（PS相当于粗糙地掌握了正则表达式的翻译器），当然文中的正则表达式并不全是我写的，我是问了 AI，在它的答案基础上修改的，后面批量处理的代码也是先投喂给它处理一个的情况，然后找它给我写完整的。

**Comment by hongtaoh** [comment]
你这网页抓取用得很溜啊

**Reply by earfanfan** [reply]
嘶……我本来想谦虚一小丢丢的，不过一想到昨天还吐槽别人过度谦虚来着，可是还是忍不住想说，其实也还有待进步啦。哎哎，不行呐，读书的时候被谦虚刻入骨髓了。

**Comment by XiangyunHuang** [comment]
厉害，我感觉你可以把区县级的统计数据都能扒拉下来了。我是在 [中国常住人口分布图（2023 年市级）](https://xiangyun.rbind.io/2025/09/china-population/)捞了各省统计年鉴，一个官网一个官网去找的。你这批量抓取挺好，要不要接着来个数据可视化呈现？

**Reply by earfanfan** [reply]
可视化呈现没咋搞，但是其实之前就写了文章的，见[看一看我国机关事业单位的养老标准](https://yuanfan.rbind.io/project/red-and-black/)，真正的结论见评论区最后一楼。

等我猴年马月支棱起来的话，其实也还是想做点可视化的，主要这事就是俺的心头好。
